galley/pkg/config/processor/transforms/serviceentry/transformer.go

```diff
@@ -0,0 +1,345 @@
+	"istio.io/istio/galley/pkg/config/processor/transforms/serviceentry/pod"
+	"istio.io/istio/galley/pkg/config/resource"
+	"istio.io/istio/galley/pkg/runtime/monitoring"
+)
+
+type transformer struct {
+	options processing.ProcessorOptions
+
+	converter *converter.Instance
+
+	services  map[resource.Name]*resource.Entry
+	endpoints map[resource.Name]*resource.Entry
+	ipToName  map[string]map[resource.Name]struct{}
+
+	podHandler  event.Handler
+	nodeHandler event.Handler
+
+	handler  event.Handler
+	statsCtx context.Context
+
+	fullSyncCtr int
+
+	// The version number for the current State of the object. Every time mcpResources or versions change,
+	// the version number also change
+	version int64
+
+	// 	mcpResources map[resource.FullName]*mcp.Resource
+}
+
+var _ event.Transformer = &transformer{}
+
+// Start implements event.Transformer
+func (t *transformer) Start() {
+	t.ipToName = make(map[string]map[resource.Name]struct{})
+	t.services = make(map[resource.Name]*resource.Entry)
+	t.endpoints = make(map[resource.Name]*resource.Entry)
+
+	podCache, cacheHandler := pod.NewCache(pod.Listener{
+		PodAdded:   t.podUpdated,
+		PodUpdated: t.podUpdated,
+		PodDeleted: t.podUpdated,
+	})
+
+	t.podHandler = cacheHandler
+	t.nodeHandler = cacheHandler
+
+	t.converter = converter.New(t.options.DomainSuffix, podCache)
+
+	statsCtx, err := tag.New(context.Background(), tag.Insert(monitoring.CollectionTag,
+		metadata.IstioNetworkingV1Alpha3SyntheticServiceentries.String()))
+	if err != nil {
+		scope.Errorf("Error creating monitoring context for counting state: %v", err)
+		statsCtx = nil
+	}
+	t.statsCtx = statsCtx
+
+	t.fullSyncCtr = len(t.Inputs())
+}
+
+// Stop implements event.Transformer
+func (t *transformer) Stop() {
+	t.ipToName = nil
+	t.services = nil
+	t.endpoints = nil
+}
+
+// Select implements event.Transformer
+func (t *transformer) Select(c collection.Name, h event.Handler) {
+	switch c {
+	case metadata.IstioNetworkingV1Alpha3SyntheticServiceentries:
+		t.handler = event.CombineHandlers(t.handler, h)
+	}
+}
+
+// Inputs implements event.Transformer
+func (t *transformer) Inputs() collection.Names {
+	return collection.Names{
+		metadata.K8SCoreV1Endpoints,
+		metadata.K8SCoreV1Nodes,
+		metadata.K8SCoreV1Pods,
+		metadata.K8SCoreV1Services,
+	}
+}
+
+// Outputs implements event.Transformer
+func (t *transformer) Outputs() collection.Names {
+	return collection.Names{
+		metadata.IstioNetworkingV1Alpha3SyntheticServiceentries,
+	}
+}
+
+// Handle implements event.Transformer
+func (t *transformer) Handle(e event.Event) {
+	switch e.Kind {
+	case event.FullSync:
+		t.fullSyncCtr--
+		if t.fullSyncCtr == 0 {
+			t.dispatch(event.FullSyncFor(t.Outputs()[0]))
+		}
+		return
+
+	case event.Reset:
+		t.dispatch(event.Event{Kind: event.Reset})
+		return
+
+	case event.Added, event.Updated, event.Deleted:
+		// fallthrough
+
+	default:
+		panic(fmt.Errorf("transformer.Handle: Unexpected event received: %v", e))
```

Comment:
Is panic appropriate here or could this occur as the result of bad external input? If can happen due to bad input, then we shouldn't panic.

Reply:
In this context, there are only a limited set of events from a closed enumeration. The default case is really a programming failure (of not handling all event kinds). I am standardizing all defaults to panic.